components/raftstore/src/store/worker/split_controller.rs

```diff
@@ -0,0 +1,537 @@
+    sample_num: usize,
+    pre_sum: &[usize],
+    mut lists: Vec<T>,
+    get_mut: F,
+) -> Vec<KeyRange>
+    where
+        F: Fn(&mut T) -> &mut Vec<KeyRange>,
+{
+    let mut rng = rand::thread_rng();
+    let mut key_ranges = vec![];
+    let high_bound = pre_sum.last().unwrap();
+    for _num in 0..sample_num {
+        let d = rng.gen_range(0, *high_bound) as usize;
+        let i = match pre_sum.binary_search(&d) {
+            Ok(i) => i,
+            Err(i) => i,
+        };
+        let list = get_mut(&mut lists[i]);
+        let j = rng.gen_range(0, list.len()) as usize;
+        key_ranges.push(list.remove(j)); // Sampling without replacement
+    }
+    key_ranges
+}
+
+// RegionInfo will maintain key_ranges with sample_num length by reservoir sampling.
+// And it will save qps num and peer.
+#[derive(Debug, Clone)]
+pub struct RegionInfo {
+    pub sample_num: usize,
+    pub qps: usize,
+    pub peer: Peer,
+    pub key_ranges: Vec<KeyRange>,
+}
+
+impl RegionInfo {
+    fn new(sample_num: usize) -> RegionInfo {
+        RegionInfo {
+            sample_num,
+            qps: 0,
+            key_ranges: Vec::with_capacity(sample_num),
+            peer: Peer::default(),
+        }
+    }
+
+    fn get_qps(&self) -> usize {
+        self.qps
+    }
+
+    fn get_key_ranges_mut(&mut self) -> &mut Vec<KeyRange> {
+        &mut self.key_ranges
+    }
+
+    fn add_key_ranges(&mut self, key_ranges: Vec<KeyRange>) {
+        for key_range in key_ranges {
+            if self.key_ranges.len() < self.sample_num {
+                self.key_ranges.push(key_range);
+            } else {
+                let i = rand::thread_rng().gen_range(0, self.qps) as usize;
+                if i < self.sample_num {
+                    self.key_ranges[i] = key_range;
+                }
+            }
+            self.qps += 1;
+        }
+    }
+
+    fn update_peer(&mut self, peer: &Peer) {
+        if self.peer != *peer {
+            self.peer = peer.clone();
+        }
+    }
+}
+
+pub struct Recorder {
+    pub detect_num: u64,
+    pub peer: Peer,
+    pub key_ranges: Vec<Vec<KeyRange>>,
+    pub times: u64,
+    pub create_time: SystemTime,
+}
+
+impl Recorder {
+    fn new(detect_num: u64) -> Recorder {
+        Recorder {
+            detect_num,
+            peer: Peer::default(),
+            key_ranges: vec![],
+            times: 0,
+            create_time: SystemTime::now(),
+        }
+    }
+
+    fn record(&mut self, key_ranges: Vec<KeyRange>) {
+        self.times += 1;
+        self.key_ranges.push(key_ranges);
+    }
+
+    fn update_peer(&mut self, peer: &Peer) {
+        if self.peer != *peer {
+            self.peer = peer.clone();
+        }
+    }
+
+    fn is_ready(&self) -> bool {
+        self.times >= self.detect_num
+    }
+
+    fn collect(&mut self, config: &SplitConfig) -> Vec<u8> {
+        let pre_sum = prefix_sum(self.key_ranges.iter(), Vec::len);
+        let key_ranges = self.key_ranges.clone();
```

Comment:
Can this clone be avoided?

Reply:
No, we get samples with equal probability without replacement, so we need to delete in line87